### ISSUE TITLE
[Discussion] How do we handle state?

### ISSUE THREAD

**Issue by sanity** [kwebio/core#16]
A problem that will be faced by most people using kweb, and for which no clear pattern or recommendations currently exist, is how to handle state.

Kweb should be particularly well suited to keeping browser-state in sync with server-state in realtime, think how Facebook updates automatically when users post comments or do various other activities.

But we need to come up with some recommendations for how to do this.  Where do we keep state?  How do we keep it in sync with persisted state (perhaps in a database)?  How do we notify the UI about changes to the state, and how do we modify state in response to UI changes?  What lessons can be learned from the JavaScript ecosystem?

I'm keeping an open mind on this for now, but here is some potentially relevant reading material (please feel free to add, comment, discuss, etc):

* https://www.safaribooksonline.com/blog/2015/10/29/react-local-component-state/

**Comment by stangls** [comment]
For better understanding, let us take a look at the following example.

Assume a user issues a post on a board.
Should the other users' browser be updated on-the-fly, i.e. the data of the post is transmitted over the websocket and the UI is updated on the client-side?
Or should the post be displayed only after the next page-refresh, i.e. HTML is generated on the server-side?

Due to the nature of the kweb, both approaches seem reasonable.
It would be nice to have both using only one implementation to render the actual HTML-code.
For this, inclusion of Kotlin2js would be a nice2have feature, right?

**Comment by sanity** [comment]
I think the desired behavior is that other user's browser are updated on-the-fly, since with kweb page refreshes should be rare, rather the page should update in realtime.

Going the kotlin2js route would be a significant change from our current approach, which is to keep most of the intelligence on the server and just issue simple relatively low-level instructions to the browser.

Some day it would be really cool to combine kotlin2js with kotlin's coroutines support to allow a single applications code to be shared between server and client with transparent communication (a little like [Opa](http://opalang.org/), but that would be very ambitious and I think would need to be a separate project to kweb.

It might be nice to rewrite kweb's bootstrap javascript in kotlin and have it compile to JavaScript, but it's not high on my todo list right now, others are welcome to tackle that though :)

**Comment by stangls** [comment]
Ok, so what if we consider a complex website with different information displayed and the state on the server is changing while the page is active in the browser.

For convenience I as a programmer would like to:

1. define different view components: i.e. supply rules for rendering
1. use those components inside my layout
1. define when each component is updated (on client-events, on state-changes and/or after some timeout)
1. have some library handle it all, that is for each component:
   * automatic re-rendering
   * transfer minimal set of changes to the client

**Comment by sanity** [comment]
I've done some work on this.

Firstly I created a way to store persistent state where you can attach listeners for any change in persistent state, see: https://github.com/sanity/shoebox

shoebox itself isn't tied to kweb in any way, but the idea is to create a "glue" project that will allow "binding" of shoebox state to UI components, similar to [this example](https://www.youtube.com/watch?v=0Q-BUldFZjA) with TornadoFX on Android.

Thoughts?
